tools/data-validator/src/main/java/data/validator/DataValidator.java

```diff
@@ -0,0 +1,95 @@
+package data.validator;
+
+import com.datastax.driver.core.*;
+
+import static data.validator.constants.Constants.*;
+
+import data.validator.cql.CQLBuilder;
+import data.validator.jobs.RowValidatorJob;
+import eu.europeana.cloud.cassandra.CassandraConnectionProvider;
+
+import javax.annotation.Resource;
+import java.util.*;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+
+/**
+ * Created by Tarek on 4/26/2017.
+ */
+
+
+public class DataValidator {
+    @Resource
+    private CassandraConnectionProvider sourceCassandraConnectionProvider;
+
+    @Resource
+    private CassandraConnectionProvider targetCassandraConnectionProvider;
+
+
+    private static final String SELECT_COLUMN_NAMES = "SELECT " + COLUMN_NAME_SELECTOR + ", " + COLUMN_INDEX_TYPE + " FROM " + SYSTEM_SCHEMA_COLUMNS_TABLE +
+            "  WHERE keyspace_name =  ?  AND columnfamily_name = ? ;";
+
+
+    public void validate(String sourceTableName, String targetTableName, int threadsCount) {
+        Session session = null;
+        Session targetSession = null;
+        ExecutorService executorService = null;
+        try {
+            executorService = Executors.newFixedThreadPool(threadsCount);
+            session = sourceCassandraConnectionProvider.getSession();
+            List<String> primaryKeys = getPrimaryKeysNames(sourceTableName, session);
+            String selectPrimaryKeysFromSourceTable = CQLBuilder.constructSelectPrimaryKeysFromSourceTable(sourceTableName, primaryKeys);
+            PreparedStatement sourceSelectStatement = session.prepare(selectPrimaryKeysFromSourceTable);
+            sourceSelectStatement.setConsistencyLevel(sourceCassandraConnectionProvider.getConsistencyLevel());
+            BoundStatement boundStatement = sourceSelectStatement.bind();
+            ResultSet rs = session.execute(boundStatement);
+            Iterator<Row> iterator = rs.iterator();
+
+            String matchCountStatementCQL = CQLBuilder.getMatchCountStatementFromTargetTable(targetTableName, primaryKeys);
+            targetSession = targetCassandraConnectionProvider.getSession();
+            PreparedStatement matchCountStatementCQLStatement = targetSession.prepare(matchCountStatementCQL);
+            matchCountStatementCQLStatement.setConsistencyLevel(targetCassandraConnectionProvider.getConsistencyLevel());
+            BoundStatement matchingBoundStatement = matchCountStatementCQLStatement.bind();
+
+            while (iterator.hasNext()) {
+                Row row = iterator.next();
+                Callable<Long> callable = new RowValidatorJob(targetSession, primaryKeys, matchingBoundStatement, row);
+                Future<Long> future = executorService.submit(callable);
```

Comment:
In my opinion when you create a separate thread for each row the cost is probably similar to the cost of running the validation of this row in the main thread. I did not test it but I would go with some solution working on a bunch of rows per thread instead of just one row.